contracts/adapters/AaveLeverageStrategyExtension.sol

```diff
@@ -0,0 +1,1273 @@
+    )
+        internal
+        view
+        returns (uint256, uint256)
+    {
+        // Calculate absolute value of difference between new and current leverage ratio
+        uint256 leverageRatioDifference = _isLever ? _newLeverageRatio.sub(_leverageInfo.currentLeverageRatio) : _leverageInfo.currentLeverageRatio.sub(_newLeverageRatio);
+
+        uint256 totalRebalanceNotional = leverageRatioDifference.preciseDiv(_leverageInfo.currentLeverageRatio).preciseMul(_leverageInfo.action.collateralBalance);
+
+        uint256 maxBorrow = _calculateMaxBorrowCollateral(_leverageInfo.action, _isLever);
+
+        uint256 chunkRebalanceNotional = Math.min(Math.min(maxBorrow, totalRebalanceNotional), _leverageInfo.twapMaxTradeSize);
+
+        return (chunkRebalanceNotional, totalRebalanceNotional);
+    }
+
+    /**
+     * Calculate the max borrow / repay amount allowed in base units for lever / delever. This is due to overcollateralization requirements on
+     * assets deposited in lending protocols for borrowing.
+     * 
+     * For lever, max borrow is calculated as:
+     * (Net borrow limit in USD - existing borrow value in USD) / collateral asset price adjusted for decimals
+     *
+     * For delever, max repay is calculated as:
+     * Collateral balance in base units * (net borrow limit in USD - existing borrow value in USD) / net borrow limit in USD
+     *
+     * Net borrow limit for levering is calculated as:
+     * The collateral value in USD * Aave collateral factor * (1 - unutilized leverage %)
+     *
+     * Net repay limit for delevering is calculated as:
+     * The collateral value in USD * Aave liquiditon threshold * (1 - unutilized leverage %)
+     *
+     * return uint256          Max borrow notional denominated in collateral asset
+     */
+    function _calculateMaxBorrowCollateral(ActionInfo memory _actionInfo, bool _isLever) internal view returns(uint256) {
+        
+        // Retrieve collateral factor and liquidation threshold for the collateral asset in precise units (1e16 = 1%)
+        ( , uint256 maxLtvRaw, uint256 liquidationThresholdRaw, , , , , , ,) = strategy.aaveProtocolDataProvider.getReserveConfigurationData(address(strategy.collateralAsset));
+
+        // Normalize LTV and liquidation threshold to precise units. LTV is measured in 4 decimals in Aave which is why we must multiply by 1e14
+        // for example ETH has an LTV value of 8000 which represents 80%
+        if (_isLever) {
+            uint256 netBorrowLimit = _actionInfo.collateralValue
+                .preciseMul(maxLtvRaw.mul(10 ** 14))
+                .preciseMul(PreciseUnitMath.preciseUnit().sub(execution.unutilizedLeveragePercentage));
+
+            return netBorrowLimit
+                .sub(_actionInfo.borrowValue)
+                .preciseDiv(_actionInfo.collateralPrice);
+        } else {
+            uint256 netRepayLimit = _actionInfo.collateralValue
+                .preciseMul(liquidationThresholdRaw.mul(10 ** 14))
```

Comment:
Why are we using `liquidationThresholdRaw` for one and `maxLtvRaw` for another? Those are different values right? It feels like we should pick one, probably the most conservative one?

Reply:
Aave allows you to withdraw up to the `liquidationThresholdRaw` and only borrow up to the `maxLtvRaw` where `liquidationThresholdRaw` > `maxLtvRaw`.

We actually want to use the higher number to delever faster. We will be repaying the borrow in the same transaction so it makes no difference

Reply:
I think this is probably ok, just want to make sure there's not a situation we can get stuck in where we're caught in between these two ratios and can't execute an action.